core/store/src/trie/mod.rs

```diff
@@ -1610,32 +1612,76 @@ impl Trie {
         }
     }
 
-    pub fn iter<'a>(&'a self) -> Result<TrieIterator<'a>, StorageError> {
-        TrieIterator::new(self, None)
+    /// Returns an iterator that can be used to traverse any range in the trie.
+    /// This only uses the on-disk trie. If memtrie iteration is desired, see
+    /// `lock_for_iter`.
+    pub fn disk_iter(&self) -> Result<DiskTrieIterator<'_>, StorageError> {
+        DiskTrieIterator::new(self, None)
     }
 
-    pub fn iter_with_max_depth<'a>(
+    pub fn disk_iter_with_max_depth<'a>(
         &'a self,
         max_depth: usize,
-    ) -> Result<TrieIterator<'a>, StorageError> {
-        TrieIterator::new(
+    ) -> Result<DiskTrieIterator<'a>, StorageError> {
+        DiskTrieIterator::new(
             self,
             Some(Box::new(move |key_nibbles: &Vec<u8>| key_nibbles.len() > max_depth)),
         )
     }
 
-    pub fn iter_with_prune_condition<'a>(
+    pub fn disk_iter_with_prune_condition<'a>(
         &'a self,
         prune_condition: Option<Box<dyn Fn(&Vec<u8>) -> bool>>,
-    ) -> Result<TrieIterator<'a>, StorageError> {
-        TrieIterator::new(self, prune_condition)
+    ) -> Result<DiskTrieIterator<'a>, StorageError> {
+        DiskTrieIterator::new(self, prune_condition)
+    }
+
+    /// Grabs a read lock on the trie, so that a memtrie iterator can be
+    /// constructed afterward. This is needed because memtries are not
+    /// thread-safe.
+    pub fn lock_for_iter(&self) -> TrieWithReadLock<'_> {
+        TrieWithReadLock { trie: self, memtries: self.memtries.as_ref().map(|m| m.read().unwrap()) }
     }
 
     pub fn get_trie_nodes_count(&self) -> TrieNodesCount {
         self.accounting_cache.borrow().get_trie_nodes_count()
     }
 }
 
+/// A wrapper around `Trie`, but holding a read lock on memtries if they are present.
+/// This is needed to construct an memtrie iterator, as memtries are not thread-safe.
+pub struct TrieWithReadLock<'a> {
+    trie: &'a Trie,
+    memtries: Option<RwLockReadGuard<'a, MemTries>>,
+}
+
+impl<'a> TrieWithReadLock<'a> {
+    /// Obtains an iterator that can be used to traverse any range in the trie.
+    /// If memtries are present, returns an iterator that traverses the memtrie.
+    /// Otherwise, it falls back to an iterator that traverses the on-disk trie.
+    pub fn iter(&self) -> Result<TrieIterator<'_>, StorageError> {
```

Comment:
With dynamic dispatch you could return `Box<dyn Iterator>` here... thus avoiding any additional types (although you would need to define a new trait that supertraits `Iterator` for `seek_prefix`)

Reply:
I like to avoid indirection if not needed, so `Box<dyn ...>` would be less appealing in my opinion, as one needs to click through cross-references in order to understand the code flow.